### DISCUSSION TITLE
Error building firmware on Mac M1

### DISCUSSION THREAD

**Discussion by mattjoyce** [shanteacontrols/OpenDeck#132] [General]
Hello,  I have installed the software according to the wiki, and the first error was an error relating to objcopy

> mattjoyce@matts-air src % make TARGET=blackpill401cc                                                                 
Toolchain.mk:47: *** Required package not found: objcopy.  Stop.

Looking on my system I have the following

> /Users/mattjoyce/gcc-arm-none-eabi-10-2020-q4-major/share/doc/gcc-arm-none-eabi/html/binutils.html/objcopy.html
/Users/mattjoyce/gcc-arm-none-eabi-10-2020-q4-major/share/doc/gcc-arm-none-eabi/man/man1/arm-none-eabi-objcopy.1
/Users/mattjoyce/gcc-arm-none-eabi-10-2020-q4-major/bin/arm-none-eabi-objcopy
/Users/mattjoyce/gcc-arm-none-eabi-10-2020-q4-major/arm-none-eabi/bin/objcopy
/Users/mattjoyce/arduino-1.8.13/Contents/Java/hardware/tools/avr/bin/avr-objcopy
/Users/mattjoyce/arduino-1.8.13/Contents/Java/hardware/tools/avr/avr/bin/objcopy

I added /Users/mattjoyce/gcc-arm-none-eabi-10-2020-q4-major/arm-none-eabi/bin/ to my path and tried again.
Quite promising, but there is still an error output.

> mattjoyce@matts-air src % make TARGET=blackpill401cc                                                             
...
Creating executable: build/app/blackpill401cc/release/blackpill401cc.elf
__TEXT  __DATA  __OBJC  others  dec     hex
error: /Library/Developer/CommandLineTools/usr/bin/size: file: ./build/app/blackpill401cc/release/blackpill401cc.elf is not an object file
make[1]: *** [build/app/blackpill401cc/release/blackpill401cc.elf] Error 1
make: *** [merged] Error 2

I get the same error for blackpill401ce, blackpill411 and opendeck2, but mega2560 works fine, as does teensy2pp.

**Comment by mattjoyce** [comment]
I tried on Ubuntu and it worked fine.  So something odd about STM32 and M1 Mac?

**Reply by paradajz** [reply]
Most likely. As I mentioned in the wiki, M1 isn't tested since I don't have one. I'll see if I can setup CI pipeline on Github for M1.

**Reply by paradajz** [reply]
Not supported: https://github.com/actions/virtual-environments/issues/2187

Sorry. Intel Mac or Ubuntu it is.

**Comment by paradajz** [comment]
Do you have `llvm-objcopy` or `gobjcopy` on you mac installed?

**Reply by mattjoyce** [reply]
actual yes, I found gobjcopy, also objcopy in the same folder v2.37
`/System/Volumes/Data/opt/homebrew/Cellar/binutils/2.37/bin/gobjcopy`

**Comment by paradajz** [comment]
I think I know what's the issue here. Homebrew has traditionally been installed to `/usr/local`. On M1 that has been changed to /opt. Now, if you've followed the building wiki to the letter, on one line there is this:

`echo 'export PATH=~/arduino-1.8.19/Contents/Java/hardware/tools/avr/bin:~/gcc-arm-none-eabi-10.3-2021.10/bin:/usr/local/opt/binutils/bin:${PATH}' >> ~/.profile`

Note the /**usr/local/opt/**. For M1 this needs to be changed. The reason for that line in the first place is because Homebrew doesn't automatically add GNU binutils to the PATH in order for GNU tools not to shadow internal mac tools. Well, except, the OpenDeck build needs GNU tools, which is why the instruction is to add them to the PATH. Since this path doesn't exist on M1, you were getting the error that objcopy doesn't exist. I assume the binutils on M1 are then in `/opt/binutils/bin` but please verify. Once you find it, just add it to the PATH and the build should work (I hope).

There was also unneeded package listed in the wiki, so you should remove it - `brew remove arm-none-eabi-gcc`.

**Reply by mattjoyce** [reply]
hi, thanks for sticking with this, and helping me.
I had a look around and it took me a while to realise mdfind does not search everywhere (sorry I'm new to MacOs)
This path is what is needed `/opt/homebrew/opt/binutils/bin/`
So I edited .profile to read
`export PATH=~/arduino-1.8.13/Contents/Java/hardware/tools/avr/bin:~/gcc-arm-none-eabi-10-2020-q4-major/bin:/opt/homebrew/opt/binutils/bin:${PATH}`

That seems to work well.

mattjoyce@matts-air src % make TARGET=blackpill401cc
Building application...
make[1]: Nothing to be done for `pre-build'.
make[1]: Nothing to be done for `binary'.
Building bootloader...
Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
Building: board/arch/arm/st/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ctlreq.c
Building: board/arch/arm/st/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_core.c
Building: board/arch/arm/st/gen/common/Middlewares/ST/STM32_USB_Device_Library/Core/Src/usbd_ioreq.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_spi.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_uart.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_exti.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_ll_usb.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_rcc_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_cortex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pwr_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_dma.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_adc.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_i2c_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_flash_ramfunc.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_tim_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_pcd_ex.c
Building: board/arch/arm/st/gen/f4/common/Drivers/STM32F4xx_HAL_Driver/Src/stm32f4xx_hal_gpio.c
Building: board/arch/arm/st/gen/f4/common/Src/system_stm32f4xx.c
Building: board/arch/arm/st/gen/f4/common/Src/usbd_conf.c
Building: board/arch/arm/st/variants/f4/Setup.cpp
Building: board/arch/arm/st/variants/f4/Flash.cpp
Building: board/arch/arm/st/variants/f4/ISR.cpp
Building: board/common/Common.cpp
Building: board/common/Bootloader.cpp
Building: board/arch/arm/common/Multiplexer.cpp
Building: board/arch/arm/common/Bootloader.cpp
Building: board/arch/arm/common/Setup.cpp
Building: board/arch/arm/st/common/ShiftRegistersWait.cpp
Building: board/arch/arm/st/common/UniqueID.cpp
Building: board/arch/arm/st/common/Setup.cpp
Building: board/arch/arm/st/common/NVM.cpp
Building: board/arch/arm/st/common/ISR.cpp
Building: board/arch/arm/st/variants/f4/stm32f401cc/Descriptors.cpp
Building: board/common/io/Stubs.cpp
Building: board/common/io/Indicators.cpp
Building: bootloader/SysExParser/SysExParser.cpp
Building: bootloader/FwSelector/FwSelector.cpp
Building: bootloader/updater/Updater.cpp
Building: bootloader/main.cpp
Building: board/common/comm/usb/descriptors/midi/Descriptors.c
Building: board/arch/arm/st/comm/usb/midi/MIDI.cpp
Creating executable: build/boot/blackpill401cc/release/blackpill401cc.elf
   text    data     bss     dec     hex filename
  19576      24   16604   36204    8d6c ./build/boot/blackpill401cc/release/blackpill401cc.elf
Building sysexgen...
Building: sysexgen/main.cpp
Creating executable: build/sysexgen/blackpill401cc/release/blackpill401cc.elf
   text    data     bss     dec     hex filename
  11869    1200       0   13069    330d ./build/sysexgen/blackpill401cc/release/blackpill401cc.elf
Building flashgen...
Building: application/database/CustomInit.cpp
Building: application/database/Database.cpp
Building: ../modules/dbms/src/LESSDB.cpp
Building: ../modules/EmuEEPROM/src/EmuEEPROM.cpp
Building: flashgen/main.cpp
Building: application/util/configurable/Configurable.cpp
Building: application/util/conversion/Conversion.cpp
Creating executable: build/flashgen/blackpill401cc/release/blackpill401cc.elf
   text    data     bss     dec     hex filename
  26113    2720    8896   37729    9361 ./build/flashgen/blackpill401cc/release/blackpill401cc.elf
Merged binary created: ./build/merged/blackpill401cc/release/blackpill401cc.hex
Creating SysEx file...
Firmware size is 173234 bytes. Generating SysEx file, please wait...
SysEx file created: ./build/merged/blackpill401cc/release/blackpill401cc.sysex


I will flash it today and test!

**Comment by paradajz** [comment]
Thanks. I've updated the instructions and added the path for M1 as well.

**Reply by mattjoyce** [reply]
To test I erased my local repo, and ran through the process from beginning to end.
No errors, but I saw my blackpillf401cc.bin was only 75kb, the one in your releases is 234kb
I tried mega2560, and that also produced a much smaller bin, 75kb again.

So, I fired up a WLS Ubuntu Linux, and that also has the same issue, 75kb binaries.

So I think compiling on the Mac is solved, but there something else not unique to mac going on.

**Reply by mattjoyce** [reply]
I will log a separate issue.

**Reply by paradajz** [reply]
Binary in which directory?

**Reply by mattjoyce** [reply]
Ah, my mistake, I was looking in `build/app/mega2560/release` instead of `build/mergedmega2560/release`

**Reply by mattjoyce** [reply]
I should read documentation more carefully.  Apologies.
